website/hw/hw4.rst

```diff
@@ -180,103 +218,557 @@ As you are stepping through your program, you might want to be able to
+
+You can build the ``micro`` editor by going into the ``micro/`` directory in
+your repository and running the following::
+
+    cmake -B build/
+    make -C build/
+
+This will generate a ``micro`` executable inside a ``build/`` directory.
+Let's give it a try! Run the following from the ``micro/`` directory::
+
+    build/micro
+
+This will open the editor with a "blank file". You can start typing to edit
+the file, and you'll notice that you can move around with the arrow keys, use
+the Backspace key, and the Delete key. You should be able to quit the editor
+by pressing Ctrl-Q (if you modified the file, you'll have to press it three
+times to confirm you want to exit without saving).
+
+However, we have inserted a few bugs in the code, so don't be surprised if
+you do something that makes the editor crash! Please note that, if this
+happens, your terminal may be left in a seemingly unusable state (in particular,
+it will seem like you can't type anything into the terminal). When this happens,
+press Enter followed by typing the word ``reset`` followed by pressing Enter (this doesn't
+reset your computer, it just resets your terminal). You can also just close the
+terminal and open it up again.
+
+Let's say we wanted to debug this executable. We actually cannot call GDB like
+before, because the text editor needs to use the entire terminal, so it
+would be impractical to have a GDB prompt interfering with that. So, this will
+allow us to do *remote debugging*, where the program and the debugger are
+run separately (and could even be running in completely different computers,
+and communicating over the network).
+
+To do this, open two terminals (if you're logging to a CS Linux server, make
+sure you're logged into the same Linux server on both terminals). On the first one, run the following from the
+``micro/`` directory::
+
+    gdbserver :50000 build/micro
+
+The editor won't actually start just yet (this is normal!) and you'll see something
+like this::
+
+    Process build/micro created; pid = 549651
+    Listening on port 50000
+
+.. note::
+
+   If you see the following message::
+
+        Can't bind address: Address already in use.
+        Exiting
+
+   Then pick a number other than 50000 in the parameter to ``gdbserver``
+   (any number between 50000 and 60000 should work). This number is known
+   as a "port", and we will be using it to connect to the editor from GDB.
+   We have to pick a number that is unique within the machine we are running
+   on; this means that if you are running this on one of the CS Linux server,
+   and multiple people choose 50000, only one of them will actually be able
+   to run the debugger.
+
+Now, on the second terminal, run this::
+
+    gdb build/micro
+
+This will open the usual GDB prompt but, instead of typing "run", we need to
+connect to the editor that is running on the other terminal. You need
+to run the following command from the GDB prompt::
+
+    target remote :50000
```

Comment:
Perhaps add something like "where <:50000> is the port number you chose" since we said they can use anything between 50000 and 60000.